### PR TITLE
Remove memory space generalization from SharedAlloc.hpp

### PR DESCRIPTION
**Pull request by jeffmiles63**
While working on the Umpire memory space, I discovered an idiosyncrasy in the template for `SharedAllocationRecord<MemorySpace, void>`.  

There was a generic implementation that was empty in Kokkos_SharedAlloc.hpp which was predominantly used by AnonymousSpace and ScratchMemorySpace.  The problem with this is that if the specialization of the SharedAllocationRecord used anything other that a direct reference to a memory space (i.e. SharedAllocationRecord<Kokkos::SomeMemorySpace, void>) then the compiler would complain about having two specializations....

My example used something like this: 

```
template<MemorySpace>
class SharedAllocationRecord<MemorySpace, std::enable_if< 
                                         is_my_space_special<MemorySpace>::value, void>::type > { };
``` 

So the fix is to remove the template from Kokkos_SharedAlloc.hpp and to add specializations for Anonymous and ScratchMemorySpace.

### CODE REVIEW

File: core/src/Kokkos_ScratchSpace.hpp [hunk 1]
```diff
@@ -208,6 +208,16 @@ class ScratchMemorySpace {
   }
 };
 
+namespace Impl {
+
+// specialize SharedAllocationRecord -- not it is never used so it doen't need
```
Comment (dalg24) [inline core/src/Kokkos_ScratchSpace.hpp]: Typos

Comment (jeffmiles63) [reply core/src/Kokkos_ScratchSpace.hpp]: fixed

**Comment by crtrott** [comment]
Does this break remotespaces?

**Comment by jeffmiles63** [comment]
@crtrott They all appear to have specializations, so no.  I will run some tests to confirm.

**Comment by masterleinad** [comment]
Currently, `clang-format` is failing.

**Comment by dhollman** [comment]
The problem here is that the second template parameter wasn't intended to be a SFINAE parameter, but rather as part of Carter's (not widely used by anyone but him AFAIK) pattern of "inherit from a specialization of `void` rather than name the base class something else". I'm still fine with the change, but that's why you're running in to issues here.

**Comment by jeffmiles63** [comment]
Closing - will document in backend development guide how to derive redirected memory spaces...

i.e. 

```c++
template<class MemorySpace>
class SharedAllocationRecord<MyDerivedMemorySpace<MemorySpace>, void>
```

**Pull request closed without merging**